File: netgo/src/main/java/com/fungo/netgo/request/FungoApi.java
package com.fungo.netgo.request;

import com.google.gson.JsonElement;

import io.reactivex.Flowable;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Url;

/**
 * @author Pinger
 * @since 18-10-18 下午3:20
 *
 * Retrofit的Service的方法上不能带有泛型，就是不能有未知类型，这里将原来的泛型改成了JsonElement,自己主动去解析数据
 *
 */
public interface FungoApi {

    @POST("{url}")
    Flowable<JsonElement> postRequest(@Path("url") String url, @Body BaseBodyRequest params);

    @POST
    Flowable<JsonElement> postRequestWithFullUrl(@Url String url, @Body BaseBodyRequest params);

    @POST("{url}")
    Flowable<JsonElement> getRequest(@Path("url") String url);

    @GET
    Flowable<JsonElement> getRequestWithFullUrl(@Url String url);

}

File: netgo/src/main/java/com/fungo/netgo/api/FungoApi.kt
package com.fungo.netgo.api

import com.fungo.netgo.entity.BaseRequestInfo
import io.reactivex.Observable
import retrofit2.http.*

/**
 * Author  ZYH
 * Date    2018/1/24
 * Des     网络请求api
 */
interface FungoApi {

    @POST("{sourceUrl}")
    fun <T> postRequest(@Path("sourceUrl") sourceUrl: String, @Body params: BaseRequestInfo): Observable<T>

    @POST
    fun <T> postRequestWithFullUrl(@Url sourceUrl: String, @Body params: BaseRequestInfo): Observable<T>

    @POST("{sourceUrl}")
    fun <T> getRequest(@Path("sourceUrl") sourceUrl: String): Observable<T>

    @GET
    fun <T> getRequestWithFullUrl(@Url sourceUrl: String): Observable<T>
}
